File: treehouse-algorithms-and-datastructures/Algorithms/Binary_Search/rec_binary_search.py
def recursive_binary_search(list, target):
  if len(list) == 0:
    return False
  else:
    midpoint = (len(list))//2
    if list[midpoint] == target:
      return True
    else:
      if list[midpoint] < target:
        return recursive_binary_search(list[midpoint + 1:], target)
      else:
        return recursive_binary_search(list[:midpoint], target)
      
def verify(result):
  print("Target Found: ", result)
  

numbers = [1,2,3,4,5,6,7,8]
result = recursive_binary_search(numbers, 4)
verify(result)

"""
Iterative Binary Search
"""

def binary_search(list, target):
    first = 0
    last = len(list) - 1

    while first <= last:
        midpoint = (first + last)//2
        if list[midpoint] == target:
            return midpoint
        elif list[midpoint] < target:
            first = midpoint+1
        else:
            last = midpoint-1
    return -1

File: treehouse-algorithms-and-datastructures/Algorithms/Binary_Search/recBinarySearch.js
const recursiveBinarySearch = (list, target) => {
    if(list.length === 0){
        return false
    } else {
        let midpoint = Math.floor(list.length / 2)
        if(list[midpoint] === target){
            return true
        } else {
            if(list[midpoint] < target){
                return recursiveBinarySearch(list.slice(midpoint + 1), target)
            } else {
                return recursiveBinarySearch(list.slice(0, midpoint), target)
            }
        }
    }
}

console.log(recursiveBinarySearch([1,2,3,4,5,6,7,8,9,10], 3))

const iterativeBinarySearch = (list, target) => {
    let first = 0
    let last = list.length - 1
    while(first <= last){
        midpoint = Math.floor((first + last) / 2)
        if(list[midpoint] == target){
            return midpoint
        } else if(list[midpoint] < target){
            first = midpoint + 1
        } else {
            last = midpoint - 1
        }
    }
    return -1
}

